eth/db/storage.py

```diff
@@ -0,0 +1,288 @@
+
+        self._trie_nodes_batch.clear()
+
+
+
+class StorageDB(BaseStorageDB):
+    """
+    Caches lookups and batches writes to storage. One key goal is to only generate
+    storage roots at the end of the block (at least post-Byzantium).
+    """
+    logger = logging.getLogger("eth.db.storage.StorageDB")
+
+    def __init__(self, db: BaseAtomicDB, account_db: BaseAccountDB) -> None:
+        self._account_db = account_db
+        self._raw_store_db = db
+        self._storage_lookup = StorageLookup(self._raw_store_db, account_db)
+        self._transient_batch_db = BatchDB(self._storage_lookup)
+        self._storage_cache = CacheDB(self._transient_batch_db)
+        self._journal_storage = JournalDB(self._storage_cache)
+
+    def _global_storage_key(self, address: Address, slot: int) -> bytes:
+        return address + int_to_big_endian(slot)
+
+    def get(self, address: Address, slot: int, from_journal: bool=True) -> int:
+        key = self._global_storage_key(address, slot)
+        lookup_db = self._journal_storage if from_journal else self._storage_cache
+        try:
+            encoded_value = lookup_db[key]
+        except KeyError:
+            return 0
+
+        if encoded_value == b'':
+            return 0
+        else:
+            return rlp.decode(encoded_value, sedes=rlp.sedes.big_endian_int)
+
+    def set(self, address: Address, slot: int, value: int) -> None:
+        self.logger.debug2("Setting decoded storage in account 0x%s & slot %s to value %s", address.hex(), slot, value)
+        key = self._global_storage_key(address, slot)
+        if value:
+            self._journal_storage[key] = rlp.encode(value)
+        else:
+            del self._journal_storage[key]
+
+    def delete(self, address: Address) -> None:
+        diff = self._journal_storage.diff()
+        for key in diff.new_keys():
+            if key.startswith(address):
+                self.logger.debug2("deleting item in storage: %r, matching account 0x%s", key, address.hex())
+                del self._journal_storage[key]
+            else:
+                self.logger.debug2("skipping deleting item in storage: %r, does not match account 0x%s", key, address.hex())
+        self._storage_cache.reset_cache()
+
+    def record(self) -> UUID:
+        c= self._journal_storage.record()
+        self.logger.debug2('save checkpoint %r', c)
+        return c
+
+    def discard(self, changeset: UUID) -> None:
+        self.logger.debug2('discard checkpoint %r', changeset)
+        self._journal_storage.discard(changeset)
+        self._storage_cache.reset_cache()
+
+    def commit(self, changeset: UUID) -> None:
+        self.logger.debug2('commit checkpoint %r', changeset)
+        self._journal_storage.commit(changeset)
+
+    def make_storage_roots(self) -> Tuple[Tuple[Address, Hash32], ...]:
+        """
+        Force calculation of all storage roots
+
+        MUST immediately update account storage roots for reads to be consistent
```

Comment:
Probably doesn't need to be done in this PR but this seems like a place for future optimization since we shouldn't actually need the storage root for each account to do EVM execution right?  Just an implementation detail.  In theory we should be able to use a simple mapping of `slot -> value` during execution and then write it to a trie once execution finishes, we'd just need a way to maintain that simplistic data view of account storage during the execution lifetime.

Reply:
Sorry, I know this is hard to follow without some docs. I'll write it up in the docstring before I remove the `[WIP]`.

The short answer is: that's what it's doing! :)  All the changes are in a slot/val uncommitted mapping until the storage roots are needed, when it's pushed into the trie.

This particular method is only called at the specific moments when we *do* need to calculate the state root (end-of-transaction for pre-Byzantium, or end-of-block afterward).